pkg/pgmodel/pgx.go

```diff
@@ -557,25 +591,38 @@ func (h *insertHandler) flush() {
 	if !h.hasPendingReqs() {
 		return
 	}
-	h.flushPending(h.pending)
+	h.flushPending()
 }
 
-func (h *insertHandler) flushPending(pending *pendingBuffer) {
-	err := func() error {
-		_, err := h.setSeriesIds(pending.batch.sampleInfos)
-		if err != nil {
-			return err
-		}
+func (h *insertHandler) flushPending() {
+	_, err := h.setSeriesIds(h.pending.batch.sampleInfos)
+	if err != nil {
+		h.pending.finish(err)
+		return
+	}
+
+	h.toCopiers <- copyRequest{h.pending, h.metricTableName}
+	h.pending = pendingBuffers.Get().(*pendingBuffer)
+}
 
-		_, err = h.conn.CopyFrom(
+func runCopyFrom(conn pgxConn, in chan copyRequest) {
+	for {
+		req, ok := <-in
+		if !ok {
+			return
+		}
+		_, err := conn.CopyFrom(
 			context.Background(),
-			pgx.Identifier{dataSchema, h.metricTableName},
+			pgx.Identifier{dataSchema, req.table},
 			copyColumns,
-			&pending.batch,
+			&req.data.batch,
 		)
-		return err
-	}()
+		req.data.finish(err)
+		pendingBuffers.Put(req.data)
```

Comment:
let's make a proper function for this line using our FinishXXX naming convention

Reply:
@cevian I really don't think it should be: it should never be used outside this file, and it's just adding a gratuitous layer of indirection. If we did more work than just `Put`ing in `pendingBuffers` that would be one thing, but `data.finish` is _not_ that (and, incidentally, is now renamed to `reportResults`).